### PR TITLE
#161342851 List All Non-admin Users

### PR DESCRIPTION
**Pull request by TheoOkafor**
#### What does this PR do?
- Create List-all-users functionality

#### Description of Task to be completed?
- Refactor the signup route
- Check JWT authentication
- Create the list all user controller and route

#### How should this be manually tested?
- git clone `https://github.com/andela/haven-ah-backend.git`.
- Install `node` if not installed.
- ensure that PostgreSQL is running on your local machine.
- run `npm install` to install the npm packages.
- run `npm test` to run the unit and integration tests.
To test on postman
- run `npm run start-dev`
- On postman run GET `localhost:5000/api/users` (set JWT token as`x-access-token` in header).

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
#161342851

#### Screenshots (if appropriate)
Good Response:
screen shot 2018-10-28 at 1 20 53 pm

Bad Token Response:
screen shot 2018-10-28 at 1 21 51 pm

When token is not provided:
screen shot 2018-10-28 at 1 22 20 pm

### CODE REVIEW

**Comment by PriscillaSam** [comment]
@TheoOkafor I don't think you need a token to get all users.

File: server/controllers/user.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+   * @param {object} request
+   * @param {object} response
+   * @returns {object} The JSON response to the user.
+   */
+  static async listAll(request, response) {
+    const allUsers = await userRepo.getAllUsers();
+    if (allUsers === null) {
+      return badHttpResponse(
+        response,
+        404,
```
Comment (PriscillaSam) [inline server/controllers/user.js]: When users are not found, I don't think it is a bad request. It should be 200 but maybe with a message that says there are no users yet or something similar.

**Comment by ucheg6** [comment]
Great job theodore... This looks good

File: server/controllers/user.js (hunk 1 above)
Comment (TheoOkafor) [reply server/controllers/user.js]: Okay. I see what you mean.
It's not a bad request per se.

File: server/config/config.js [hunk 2]
```diff
@@ -28,6 +28,7 @@ const config = {
     PORT: process.env.PORT,
     secret: process.env.JWT_SECRET,
   },
+  secret: process.env.DB_PASSWORD,
```
Comment (seyi-adeleke) [inline server/config/config.js]: please remove this

File: server/controllers/user.js [hunk 3]
```diff
@@ -46,7 +45,31 @@ class User {
...
+   * @param {object} request
+   * @param {object} response
+   * @returns {object} The JSON response to the user.
+   */
+  static async listAll(request, response) {
+    const allUsers = await userRepo.getAllUsers();
+    if (allUsers === null) {
+      return goodHttpResponse(
+        response,
+        404,
```
Comment (seyi-adeleke) [inline server/controllers/user.js]: is 404 the correct status code to use here?

Comment (seyi-adeleke) [reply server/controllers/user.js]: `goodHttpResponse ` takes a response obj, a status code , a message and an object as the last parameter. Lets stick to that please

File: server/middlewares/checkAuth.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import jwt from 'jsonwebtoken';
+import config from '../config/config';
+import { badHttpResponse } from '../utilities/httpResponse';
+
+const env = process.env.NODE_ENV || 'development';
```
Comment (seyi-adeleke) [inline server/middlewares/checkAuth.js]: seen this duplicated in a number of places. Can we make it a helper? `getCurrentEnv()`

File: server/repository/userRepository.js [hunk 5]
```diff
@@ -21,19 +21,67 @@ class UserRepository {
   /**
  * Finds a user by email
  * @param {string} email Email to search by
+ * @param {string} username username
  * @returns {object | null} User object or null if user is not found
  */
-  static async getUserByEmail(email) {
-    const user = await User.findOne({
-      where: { email },
+  static async getUserByCredentials(email, username) {
```
Comment (seyi-adeleke) [inline server/repository/userRepository.js:43]: Why was this change required?

File: server/repository/userRepository.js [hunk 6]
```diff
@@ -21,19 +21,67 @@ class UserRepository {
...
+          'createdAt',
+          'updatedAt',
+          'deletedAt',
+          'isDeleted',
+          'password'
+        ],
       }
     });
-    if (!user) return null;
+    if (user.length <= 0) return null;
```
Comment (seyi-adeleke) [inline server/repository/userRepository.js]: lets be consistent with our `If` statements please! No one-liners

**Review by seyi-adeleke — CHANGES_REQUESTED** [review]

File: server/controllers/user.js (hunk 3 above)
Comment (TheoOkafor) [reply server/controllers/user.js]: > is 404 the correct status code to use here?

Yes, it is.

File: server/middlewares/checkAuth.js (hunk 4 above)
Comment (TheoOkafor) [reply server/middlewares/checkAuth.js]: Okay

File: server/repository/userRepository.js (hunk 5 above)
Comment (TheoOkafor) [reply server/repository/userRepository.js:43]: We decided to check both email and username when a user is signing up. This is because we don't want two users to have the same username. Or a user to sign up again using the same username and/or email.

File: server/repository/userRepository.js (hunk 6 above)
Comment (TheoOkafor) [reply server/repository/userRepository.js]: Okay.

Comment (TheoOkafor) [reply server/repository/userRepository.js]: I will use a ternary operator instead

File: server/controllers/user.js (hunk 3 above)
Comment (seyi-adeleke) [reply server/controllers/user.js]: @TheoOkafor 404 is not the correct status code to use in this situation. This is a valid request

Comment (seyi-adeleke) [reply server/controllers/user.js]: @PriscillaSam gave you the same feedback 🤦‍♂️

**Review by seyi-adeleke — CHANGES_REQUESTED** [review]

File: server/controllers/user.js (hunk 3 above)
Comment (TheoOkafor) [reply server/controllers/user.js]: My bad.

**Review by seyi-adeleke — APPROVED** [review]

File: server/repository/userRepository.js (hunk 6 above)
Comment (seyi-adeleke) [reply server/repository/userRepository.js]: 🤦‍♂️

**Review by seyi-adeleke — APPROVED** [review]
